orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/DetermineRollbackCandidatesTask.java

```diff
@@ -105,172 +105,125 @@ public long getTimeout() {
-      } catch (Exception e) {
-        logger.warn(
-            "Failed to fetch server group, retrying! (account: {}, region: {}, serverGroup: {})",
+    AtomicReference<Moniker> moniker =
+        populateMonikerWithServerGroupInfo(
+            stageData.moniker,
             stageData.credentials,
             stageData.regions.get(0),
-            stageData.serverGroup,
-            e);
-        return TaskResult.RUNNING;
-      }
+            stageData.serverGroup);
+    if (moniker == null) {
+      return TaskResult.RUNNING;
     }
 
-    try {
-      cluster =
-          retrySupport.retry(
-              () ->
-                  fetchCluster(
-                      moniker.get().getApp(),
-                      stageData.credentials,
-                      moniker.get().getCluster(),
-                      stageData.cloudProvider),
-              5,
-              1000,
-              false);
-    } catch (Exception e) {
-      logger.warn(
-          "Failed to fetch cluster, retrying! (application: {}, account: {}, cluster: {}, cloudProvider: {})",
-          moniker.get().getApp(),
-          stageData.credentials,
-          moniker.get().getCluster(),
-          stageData.cloudProvider,
-          e);
+    List<ServerGroup> serverGroups =
+        getServerGroups(moniker, stageData.credentials, stageData.cloudProvider);
+    if (serverGroups == null) {
       return TaskResult.RUNNING;
     }
 
-    List<ServerGroup> serverGroups =
-        objectMapper.convertValue(
-            cluster.get("serverGroups"), new TypeReference<List<ServerGroup>>() {});
-    serverGroups.sort(Comparator.comparing((ServerGroup o) -> o.createdTime).reversed());
+    return determineRollbackCandidates(stageData, moniker.get().getCluster(), serverGroups);
+  }
+
+  private TaskResult determineRollbackCandidates(
+      StageData stageData, String cluster, List<ServerGroup> serverGroups) {
 
     List<Map> imagesToRestore = new ArrayList<>();
+    Map<String, String> rollbackTypes = new HashMap<>();
+    Map<String, Map> rollbackContexts = new HashMap<>();
+
     for (String region : stageData.regions) {
       List<ServerGroup> allServerGroupsInRegion =
           serverGroups.stream()
               .filter(s -> region.equalsIgnoreCase(s.region))
               .collect(Collectors.toList());
 
-      if (allServerGroupsInRegion.size() < 2) {
-        // need at least one server group to rollback from, and one to rollback to!
-        logger.warn(
-            "Not enough server groups in cluster {} and region {} to perform a rollback. Skipping this region.",
-            moniker.get().getCluster(),
-            region);
+      if (!isRollbackPossible(allServerGroupsInRegion, cluster, region)) {
         continue;
       }
 
       List<ServerGroup> enabledServerGroupsInRegion =
           allServerGroupsInRegion.stream()
-              .filter(s -> s.disabled == null || !s.disabled)
+              .filter(DetermineRollbackCandidatesTask::isEnabled)
```

Comment:
nit: i would rename this to `isServerGroupEnabled` otherwise it feels like we are talking about the task being enabled

Reply:
Fixed in 3eeb3c326591ef58dd0f16b7676e11260d6cd600